### DISCUSSION TITLE
[Question][Hw4] Minor questions

### DISCUSSION THREAD

**Discussion by ShinWonho** [prosyslab-classroom/cs524-program-analysis#378] [Q&A] [answered]
There are some things confusing for me, so I want to clarify them.

1. Abstract semantics of division
    - `The output of the abstract semantics of div is defined as top when the divisor can be potentially zero.`
    - With this instruction, `Bottom` divided by potentially zero is `Top` or `Bottom`?
    - It can be viewed as `Top` because of instruction above, and it can be also viewed as `Bottom` because no division is performed actually.
2. Abstract semantics of comparing operations
    - I think that result of comparing operation is boolean but it is value(integer) in homework.pdf.
    - What is the meaning of this?
3. Can I assume the division is integer division?

Thank you.

**Comment by KihongHeo** [comment, marked as answer]
Hi,

1-1. `The output of the abstract semantics of div is defined as top when the divisor can be potentially zero.`
This design choice is for practical uses of analyzers. For example,
```c
x = input();
y = 0;
x = x / y;   // L1
x = x / y;   // L2
```
Suppose we define `X / 0` is bottom. It is theoretically ok. But the analyzer can catch the bug at L1 only. It is also theoretically ok because L2 will be never executed. However, here we want to report both of them. So, it is top in our design. Of course this is theoretically correct, because Bot <= Top.

1-2. `Bot / 0`
Think about when it would happen. Also similarly, both Top or Bottom would be a reasonable choice. However, TAs will not test your code that is sensitive to this choice.

2-1. Let us assume boolean values are 0 and 1, for simplicity.

2-2. `What is the meaning of this?` <= What is the meaning of this?

3. Yes. Consider only integers.

**Comment by ShinWonho** [comment]
Oh, actually each question is not divided by subquestions. Sorry for the confusion...
Anyway, thank you for answering!
